crates/bevy_ecs/src/system/query.rs

```diff
@@ -11,102 +11,171 @@ use bevy_tasks::TaskPool;
 use std::{any::TypeId, fmt::Debug};
 use thiserror::Error;
 
-/// Provides scoped access to a [`World`] according to a given [`WorldQuery`] and query filter.
+/// Provides scoped access to a [`World`].
 ///
 /// Queries are a powerful tool enabling the programmer to iterate over entities and their components
 /// as well as filtering them on certain conditions.
 ///
-/// # Query Building Primer
+/// Query functionality is based on the [`WorldQuery`] trait. Both tuples of components
+/// (up to 16 elements) and query filters implement this trait.
 ///
-/// ### Basic Component Access
+/// `Query` accepts two type parameters:
 ///
-/// A basic query looks like `Query<&UnitHealth>` and all it does is grant immutable access to all
-/// `UnitHealth` components. Similarly using `&mut UnitHealth` instead grants mutable access instead.
+/// ```notrust
+/// Query</*data access*/, /*query filter (optional)*/>
+/// ```
+///
+/// # Usage as system parameter
+///
+/// ## Immutable component access
 ///
-/// The main way to access the components of a query is through the [`Query::iter`] and [`Query::iter_mut`]
-/// functions which return a [`QueryIter`] to iterate over:
+/// The following example defines a query that gives an iterator over `(&ComponentA, &ComponentB)`
+/// tuples, where `ComponentA` and `ComponentB` belong to the same entity. Accessing components
+/// immutably helps system parallelization.
 ///
 /// ```
 /// # use bevy_ecs::system::IntoSystem;
 /// # use bevy_ecs::system::Query;
-/// struct UnitHealth(pub u32);
-/// fn system(query: Query<&UnitHealth>) {
-///     for UnitHealth(health) in query.iter() {
-///         println!("We got {} health points left!", health);
-///     }
-/// }
+/// # struct ComponentA;
+/// # struct ComponentB;
+/// # fn system(
+/// query: Query<(&ComponentA, &ComponentB)>
+/// # ) {}
 /// # system.system();
 /// ```
 ///
-/// ### Multiple Component Access
+/// ## Mutable component access
 ///
-/// Instead of asking for just one component like before we can build a query that queries for multiple
-/// components with the help of tuples,`Query<(&Shape, &Color, &mut Size)>`. This query retrieves
-/// immutable references to the `Shape` and `Color` component and a mutable reference to the `Size`
-/// component.
+/// The following example similar to the previous one, with the exception of `ComponentA`
+/// being accessed mutably here. Both mutable and immutable accesses are allowed in the same
+/// query. Two systems cannot be executed in parallel if both access a certain component and
```

Comment:
This is not quite correct. If we have a guarantee that the archetypes accessed are disjoint they can run in parallel.

Reply:
I completed the sentence to add the condition. Is it correct now?

Reply:
Nope: system scheduling logic is independent of component storage <3 This is why "archetypes" still exist for sparse set component storage, and why Cart needed to make up "table-based storage type" as a new distinguishing term ;)

Reply:
Basically, the way Bevy uses "archetype" is to define a "set of components" that an entity could have.

"at least one of the accesses is mutable, unless Bevy can verify that no entity could be found in both queries"

is the logic that completes this sentence. Right now, `Without` filters are the only good way to create such a contradiction, but archetype invariants (#1481) would give us much more powerful tools to define those sorts of things up front to declare "common sense" rules like "cameras are never players", enabling more parallelism.

Reply:
Thanks for the info 😄

I updated the description with your suggestion. I feel like that the following phrase can be sligtly changed:

> unless **Bevy** can verify that no entity could be found in both queries

I think the use of the noun *Bevy* doesn't give any idea of what ECS district is responsible for that check. Would *schedule* be more appropriate?

Reply:
I think *scheduler* is even more appropriate here.